api/auth/authproviders.py

```diff
@@ -0,0 +1,246 @@
+import datetime
+import requests
+import json
+import urllib
+import urlparse
+
+from . import APIAuthProviderException, APIUnknownUserException
+from .. import config, util
+
+log = config.log
+
+class AuthProvider(object):
+    """
+    This class provides access to mongodb collection elements (called containers).
+    It is used by ContainerHandler istances for get, create, update and delete operations on containers.
+    Examples: projects, sessions, acquisitions and collections
+    """
+
+    def __init__(self, auth_type):
+        self.auth_type = auth_type
+        try:
+            self.config =  config.get_auth(auth_type)
+        except KeyError:
+            raise NotImplementedError('Auth type {} is not supported by this instance'.format(auth_type))
+
+    @staticmethod
+    def factory(auth_type):
+        """
+        Factory method to aid in the creation of an AuthProvider instance
+        when auth_type is dynamic.
+        """
+        if auth_type in AuthProviders:
+            provider_class = AuthProviders[auth_type]
+            return provider_class()
+        else:
+            raise NotImplementedError('Auth type {} is not supported'.format(auth_type))
+
+    def validate_code(self, code, **kwargs):
+        raise NotImplementedError
+
+    def ensure_user_exists(self, uid):
+        user = config.db.users.find_one({'_id': uid})
+        if not user:
+            raise APIUnknownUserException('User {} does not exist.'.format(uid))
+        if user.get('disabled', False) is True:
+            raise APIUnknownUserException('User {} is disabled'.format(uid))
+
+    def set_user_gravatar(self, uid, email):
+        if email and uid:
+            gravatar = util.resolve_gravatar(email)
+            if gravatar is not None:
+                timestamp = datetime.datetime.utcnow()
+                # Update the user's gravatar if it has changed.
+                config.db.users.update_one({'_id': uid, 'avatars.gravatar': {'$ne': gravatar}}, {'$set':{'avatars.gravatar': gravatar, 'modified': timestamp}})
+
+
+
+class JWTAuthProvider(AuthProvider):
```

Comment:
It's for LDAP. We use a third party service that shuttles username/password to an LDAP server and creates a JWT if successful. I was considering that other authtypes that don't follow OAuth2 could use the same mechanism so the API doesn't have to deal with username/password flows. In that case they would also use this JWTAuthType provider.